File: backend/src/core/TuColmadoRD.Core.Domain/Base/Result/OperationResult.cs
using System;

namespace TuColmadoRD.Core.Domain.Base.Result
{
    public class OperationResult<TResult, TError>
    {
        private readonly TResult? _result;
        private readonly TError? _error;

        public bool IsGood { get; }

        public bool TryGetResult(out TResult? result)
        {
            result = _result;
            return IsGood;
        }

        public bool TryGetError(out TError? error)
        {
            error = _error;
            return !IsGood;
        }

        public TResult Result => IsGood
            ? _result!
            : throw new InvalidOperationException($"Operation failed. Error: {_error}");

        public TError Error => !IsGood
            ? _error!
            : throw new InvalidOperationException("Cannot access error of a successful operation.");

        private OperationResult(bool isGood, TResult? result, TError? error)
        {
            IsGood = isGood;
            _result = result;
            _error = error;
        }

        public static OperationResult<TResult, TError> Good(TResult? result) =>
            new(true, result, default);

        public static OperationResult<TResult, TError> Bad(TError? error) =>
            new(false, default, error);

        public TOut Match<TOut>(
            Func<TResult?, TOut> onGood,
            Func<TError?, TOut> onBad) =>
            IsGood ? onGood(_result) : onBad(_error);

        public void Match(
            Action<TResult?> onGood,
            Action<TError?> onBad)
        {
            if (IsGood) onGood(_result);
            else onBad(_error);
        }

        public OperationResult<TNewResult, TError> Map<TNewResult>(
            Func<TResult?, TNewResult> mapper) =>
            IsGood
                ? OperationResult<TNewResult, TError>.Good(mapper(_result))
                : OperationResult<TNewResult, TError>.Bad(_error);

        public OperationResult<TNewResult, TError> Bind<TNewResult>(
            Func<TResult?, OperationResult<TNewResult, TError>> binder) =>
            IsGood
                ? binder(_result)
                : OperationResult<TNewResult, TError>.Bad(_error);

        public async Task<OperationResult<TNewResult, TError>> MapAsync<TNewResult>(
            Func<TResult?, Task<TNewResult>> mapper) =>
            IsGood
                ? OperationResult<TNewResult, TError>.Good(await mapper(_result))
                : OperationResult<TNewResult, TError>.Bad(_error);

        public async Task<OperationResult<TNewResult, TError>> BindAsync<TNewResult>(
            Func<TResult?, Task<OperationResult<TNewResult, TError>>> binder) =>
            IsGood
                ? await binder(_result)
                : OperationResult<TNewResult, TError>.Bad(_error);

        public OperationResult<TResult, TError> Tap(Action<TResult?> onGood)
        {
            if (IsGood) onGood(_result);
            return this;
        }
    }
}

File: notification-service/src/domain/result/operation-result.ts
export class OperationResult<T, E> {
  private readonly _ok: boolean;
  private readonly _value: T | undefined;
  private readonly _error: E | undefined;

  private constructor(ok: boolean, value?: T, error?: E) {
    this._ok = ok;
    this._value = value;
    this._error = error;
  }

  static ok<T, E>(value: T): OperationResult<T, E> {
    return new OperationResult<T, E>(true, value, undefined);
  }

  static fail<T, E>(error: E): OperationResult<T, E> {
    return new OperationResult<T, E>(false, undefined, error);
  }

  get isOk(): boolean {
    return this._ok;
  }

  get value(): T {
    if (!this._ok) throw new TypeError('Cannot read value of a failed OperationResult.');
    return this._value as T;
  }

  get error(): E {
    if (this._ok) throw new TypeError('Cannot read error of a successful OperationResult.');
    return this._error as E;
  }
}
